h2o-core/src/main/java/water/fvec/persist/FramePersist.java

```diff
@@ -0,0 +1,189 @@
+
+import jsr166y.CountedCompleter;
+import water.*;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.util.FileUtils;
+
+import java.net.URI;
+
+import static water.fvec.persist.PersistUtils.*;
+
+public class FramePersist {
+
+    static {
+        TypeMap.onIce(FrameMeta.class.getName());
+    }
+
+    private final Frame frame;
+
+    public FramePersist(Frame frame) {
+        this.frame = frame;
+    }
+
+    private static class FrameMeta extends Iced<FrameMeta> {
+        Key<Frame> key;
+        String[] names;
+        Vec[] vecs;
+        long[] espc;
+        int numNodes;
+
+        FrameMeta(Frame f) {
+            key = f._key;
+            names = f.names();
+            vecs = f.vecs();
+            espc = f.anyVec().espc();
+            numNodes = H2O.CLOUD.size();
+        }
+    }
+
+    private static URI getMetaUri(Key key, String dest) {
+        return FileUtils.getURI(dest + "/" + key);
+    }
+
+    private static URI getDataUri(String metaUri, int cidx) {
+        return FileUtils.getURI(metaUri + "_n" + H2O.SELF.index() + "_c" + cidx);
+    }
+
+    public Job<Frame> saveTo(String uri, boolean overwrite) {
+        URI metaUri = getMetaUri(frame._key, sanitizeUri(uri));
+        if (exists(metaUri) && !overwrite) {
+            throw new IllegalArgumentException("File already exists at " + metaUri);
+        }
+        write(metaUri, ab -> {
+            ab.put(new FrameMeta(frame));
+        });
+        Job<Frame> job = new Job<>(frame._key, "water.fvec.Frame", "Save frame");
+        return job.start(new SaveFrameDriver(job, frame, metaUri), 1);
+    }
+
+    public static class SaveFrameDriver extends H2O.H2OCountedCompleter<LoadFrameDriver> {
+
+        private final Job<Frame> job;
+        private final Frame frame;
+        private final URI metaUri;
+
+        public SaveFrameDriver(
+            Job<Frame> job, Frame frame,
+            URI metaUri
+        ) {
+            this.job = job;
+            this.frame = frame;
+            this.metaUri = metaUri;
+        }
+
+        @Override
+        public void compute2() {
+            frame.read_lock(job._key);
+            new SaveChunksTask(metaUri.toString()).doAll(frame).join();
+            tryComplete();
+        }
+
+        @Override
+        public void onCompletion(CountedCompleter caller) {
+            frame.unlock(job);
+        }
+
+        @Override
+        public boolean onExceptionalCompletion(Throwable t, CountedCompleter caller) {
+            frame.unlock(job);
+            return super.onExceptionalCompletion(t, caller);
+        }
+    }
+
+    static class SaveChunksTask extends MRTask<SaveChunksTask> {
+
+        private final String metaUri;
+
+        SaveChunksTask(String metaUri) {
+            this.metaUri = metaUri;
+        }
+
+        @Override
+        public void map(Chunk[] cs) {
+            PersistUtils.write(getDataUri(metaUri, cs[0].cidx()), ab -> writeChunks(ab, cs));
```

Comment:
the progress of the job is never incremented

Reply:
fixed